### PR TITLE
Replace deprecated GHA ::set-output syntax

### LINKED ISSUE #588: Replace deprecated GHA ::set-ouput syntax
Looks like we will need to update the way we output info from jobs in github actions: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

TL;DR - use this syntax instead:

```
echo "{name}={value}" >> $GITHUB_OUTPUT
```

e.g.
```
- name: Set color
  id: random-color-generator
  run: echo "SELECTED_COLOR=green" >> $GITHUB_OUTPUT
- name: Get color
  run: echo "The selected color is ${{ steps.random-color-generator.outputs.SELECTED_COLOR }}"
```

### PR DESCRIPTION
**Pull request by phil9909**
## Summary

Use the new way of producing outputs in a github action as the old way is deprecated and will be removed (31st May 2023).

See https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

If this PR is to big to review, I can break it down into smaller PRs. 

## Use Cases

Fixes #588

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: builder/.github/workflows/create-release.yml
Comment (robdimsdale) [inline builder/.github/workflows/create-release.yml:115]: I assume this whitespace change is still valid yaml - is the indentation change a problem?

File: actions/release/create/entrypoint/main_test.go
Comment (robdimsdale) [inline actions/release/create/entrypoint/main_test.go:153]: I think we can replace this with `os.RemoveAll` and drop the `if` condition. That's analogous to `rm -f` - it will still succeed if the file is not present.

File: actions/release/create/entrypoint/main_test.go
Comment (robdimsdale) [inline actions/release/create/entrypoint/main_test.go:38]: I see that we changed the `main_test.go` for the `release/create` action, but I don't see any changes to the `main.go` file for the `release/create` action itself.

Am I missing something?

File: actions/release/reset-draft/entrypoint/main.go
Comment (robdimsdale) [inline actions/release/reset-draft/entrypoint/main.go:98]: Similar to above - I think this error message should read: `GITHUB_OUTPUT is not set`

Comment (robdimsdale) [reply actions/release/reset-draft/entrypoint/main.go:98]: Also I'm probably fine not checking this `ok` condition - GitHub guarantees that this environment variable is set, and even if it's not set we will just fail on the next line, albeit with a less helpful error message.

I'm happy to leave this check in the code, but I'd be happy to see it removed either. I'm just pointing out the user experience is pretty much identical either way.

File: actions/release/reset-draft/entrypoint/main_test.go
Comment (robdimsdale) [inline actions/release/reset-draft/entrypoint/main_test.go:135]: As above, can we use `os.RemoveAll` and drop the `if` condition?

Comment (robdimsdale) [reply actions/release/reset-draft/entrypoint/main_test.go:135]: Also, you could use [`t.TempDir`](https://pkg.go.dev/testing#T.TempDir) with a `filepath.Join` - that will sidestep the requirement to clean up after each test.

File: actions/stack/diff-package-receipts/entrypoint/main.go
Comment (robdimsdale) [inline actions/stack/diff-package-receipts/entrypoint/main.go:147]: Another instance where the error message should say: `GITHUB_OUTPUT`

File: actions/stack/get-usns/entrypoint/main.go
Comment (robdimsdale) [inline actions/stack/get-usns/entrypoint/main.go:131]: Another instance where the error message should say: `GITHUB_OUTPUT`

File: actions/stack/release-notes/entrypoint/main.go
Comment (robdimsdale) [inline actions/stack/release-notes/entrypoint/main.go:128]: Another instance where the error message should say: `GITHUB_OUTPUT`

File: actions/stack/release-notes/entrypoint/main.go
Comment (robdimsdale) [inline actions/stack/release-notes/entrypoint/main.go:140]: I'd also be happy to use the https://github.com/google/uuid package instead of creating our own.

I think it's slightly clearer to use the `uuid` package as it captures the intent of what we're doing (and is fewer steps so less likely to accidentally change something in the future).

But I think it's fine either way.

File: actions/tag/calculate-semver/entrypoint/main.go
Comment (robdimsdale) [inline actions/tag/calculate-semver/entrypoint/main.go:240]: Can we rename this function to more clearly capture that it is only writing tags? Initially I thought it was writing any key/value pair, but now I see it is only writing tags.

Maybe call it: `writeTagOutput`?

File: actions/tag/calculate-semver/entrypoint/main_test.go
Comment (robdimsdale) [inline actions/tag/calculate-semver/entrypoint/main_test.go:224]: As above, does `os.RemoveAll` (or `t.TempDir()`) simplify this?

File: actions/tag/calculate-semver/entrypoint/main_test.go
Comment (robdimsdale) [inline actions/tag/calculate-semver/entrypoint/main_test.go:233]: Have you tested what happens when this fails? Specifically, does the failure message get associated with the correct line?

I have vague memories of using helper functions like this resulting in incorrect line info, but that might have been specific to `ginkgo`, rather than `go test` + `spec` library.

**Review by robdimsdale — CHANGES_REQUESTED** [review]
Thank you so much for submitting this!

I agree that it's long, but I don't think we need to split it into separate commits or PRs. All the changes are part of the same logical change; it's just a change that touches a lot of files.

There's a couple of typos (and maybe a missing file?) that I think we need to fix before we can merge this, and then a couple of potential refactors that won't change the functionality but might improve the maintainability of the code.

Other than that, I think this would be good to go. I'm not a tooling maintainer so obviously we need to wait for them to approve as well. Lots of folks are out this week at Kubecon North America, so it might be a few days before it can be merged.

Out of curiosity, did you do this all by hand or did you script any of it (e.g. with `sed`)?

File: builder/.github/workflows/create-release.yml
Comment (phil9909) [reply builder/.github/workflows/create-release.yml:115]: This is actually unrelated to the change, but I couldn't resist to fix it anyway (notice the there are 3 more spaces in front of `Another` compared to `comment_body`, everywhere else "one level" is 2 spaces).

Still, it is unrelated, so I can remove it from this PR, if you prefer.

File: actions/release/create/entrypoint/main_test.go
Comment (phil9909) [reply actions/release/create/entrypoint/main_test.go:38]: changed the wrong test file, will revert this part.

File: actions/tag/calculate-semver/entrypoint/main_test.go
Comment (phil9909) [reply actions/tag/calculate-semver/entrypoint/main_test.go:233]: The line reported will be in the helper function. Actually, there is [`ExpectWithOffset`](https://onsi.github.io/gomega/#making-assertions-in-helper-functions) for this use-cases, but the offset seems to be ignored (maybe only works with `ginkgo`?)

Example output:

```
--- FAIL: TestEntrypoint (0.69s)
    --- FAIL: TestEntrypoint/calculate-semver (0.22s)
        --- FAIL: TestEntrypoint/calculate-semver/all_PRs_since_the_last_release_have_semver:patch/increments_the_patch_from_the_previous_version (0.01s)
            main_test.go:231: 
                Expected
                    <[]string | len:2, cap:2>: ["tag=1.2.4", ""]
                to contain elements
                    <[]string | len:1, cap:1>: ["tag=1.2.5"]
                the missing elements were
                    <[]string | len:1, cap:1>: ["tag=1.2.5"]
FAIL
FAIL	github.com/paketo-buildpacks/github-config/actions/tag/calculate-semver-tag/entrypoint	0.696s
FAIL
```

While the line is not to helpful, the test name is printed, which should be sufficient to find the relevant part, WDYT?
Alternatively, I could make the function return the output array and move the matcher call to to individual tests, if you prefer.

File: actions/release/reset-draft/entrypoint/main.go
Comment (phil9909) [reply actions/release/reset-draft/entrypoint/main.go]: I don't mind either, anyone else with a strong opinion on this?

File: actions/stack/release-notes/entrypoint/main.go
Comment (phil9909) [reply actions/stack/release-notes/entrypoint/main.go:140]: I don't mind either, anyone else with a strong opinion on this?

**Comment by phil9909** [comment]
> Out of curiosity, did you do this all by hand or did you script any of it (e.g. with `sed`)?

Semi-automated I'd say. Obviously some parts required special care (e.g. the go tests and the multi-line stuff), but most of it could be solved by VS Code "search and replace" which supports regex and capture groups in the replacement. I guess `sed` would have worked the same.

Search: `echo "::set-output name=([^:]+)::(.+)"`
Replacement: `echo "$1=$2" >> "$GITHUB_OUTPUT"`

File: actions/tag/calculate-semver/entrypoint/main_test.go
Comment (robdimsdale) [reply actions/tag/calculate-semver/entrypoint/main_test.go:227]: I think the `ExpectWithOffset` is what I was thinking of, and it doesn't surprise me that it's only required/only works when using Ginkgo.

I think this is fine - you can see which test is failing and what the expected/desired cases are.

File: builder/.github/workflows/create-release.yml
Comment (robdimsdale) [reply builder/.github/workflows/create-release.yml:115]: yeah, I appreciate the cleanup! I'm happy to keep it in the PR.

I just wanted to make sure that the indentation doesn't matter (which I think it doesn't because it's effectively ignored in a multi-line string).

**Review by robdimsdale — CHANGES_REQUESTED** [review]
These changes look good.

I think I only have one outstanding question - can we simplify the tests via `t.TempDir()` (or `os.RemoveAll()` if `t.TempDir()` doesn't work)?

Once we make that change, or can articulate why it doesn't work, then I'm happy to approve.

**Review by robdimsdale — DISMISSED** [review]
Looks great - thanks for working through this!

**Review by ForestEckhardt — DISMISSED** [review]

**Comment by ForestEckhardt** [comment]
@paketo-buildpacks/tooling-maintainers Given the size of this change it seems to me that getting a majority of maintainer approval on this would make the most sense. I am going to hold off merging until I see more approvals.

**Comment by phil9909** [comment]
@robdimsdale @ForestEckhardt My rebase has dismissed your approvals :disappointed: 

> I am would suggest holding off on merging until there are more approvals.

Fine for me

**Review by ForestEckhardt — APPROVED** [review]

**Review by ryanmoran — APPROVED** [review]

**Review by robdimsdale — APPROVED** [review]

**Comment by phil9909** [comment]
> getting a majority of maintainer approval

Tooling maintainers are 4 people right? So with @ryanmoran and @ForestEckhardt approving we have a 50%, which is technically not a majority.

@fg-j @sophiewigmore one of you wanna have a look at this as well?

**Review by sophiewigmore — APPROVED** [review]

**Comment by ForestEckhardt** [comment]
@phil9909 I think we have a super majority now. If we could get you to rebase this one more time so we could merge it that would be appreciated!

**Comment by phil9909** [comment]
I rebased and made sure no new `::set-ouput` usages were introduced

**Review by ForestEckhardt — APPROVED** [review]

**Comment by ForestEckhardt** [comment]
Thank you very much for this PR!

**Pull request merged**
